weld/subsystem/src/main/java/org/jboss/as/weld/WeldStartCompletionService.java

```diff
@@ -44,13 +49,34 @@
     private final InjectedValue<WeldBootstrapService> bootstrap = new InjectedValue<WeldBootstrapService>();
 
     private final ClassLoader classLoader;
+    private final List<ServiceController> serviceControllers;
+
+    private final AtomicBoolean runOnce = new AtomicBoolean();
 
-    public WeldStartCompletionService(ClassLoader classLoader) {
+    public WeldStartCompletionService(ClassLoader classLoader, List<ServiceController> serviceControllers) {
         this.classLoader = classLoader;
+        this.serviceControllers = serviceControllers;
     }
 
     @Override
     public void start(final StartContext context) throws StartException {
+        if (!runOnce.compareAndSet(false,true)) {
+            // we only execute this once, if there is a restart, WeldStartService initiates re-deploy hence we do nothing here
+            return;
+        }
+
+        StabilityMonitor monitor = new StabilityMonitor();
+        for (ServiceController controller : serviceControllers) {
+            monitor.addController(controller);
+        }
+        try {
+            monitor.awaitStability();
```

Comment:
This would need to be executed via the server service thread pool. If MSC was only running with a single thread this would cause a deadlock.

Reply:
Perhaps a better approach would be to just add a listener to each service, which decrements a count, and then complete the async start when the count hits zero, which would remove the need for a waiting thread.

Reply:
Re executor: yep, a service start/stop should never block, see also https://github.com/wildfly/wildfly/blob/master/weld/subsystem/src/main/java/org/jboss/as/weld/WeldBootstrapService.java#L146-L172

Re listener approach: I'm not sure I'm following. How would this remove the need for a waiting thread? It would be still blocking, or?